src/puppetlabs/puppetdb/cli/generate.clj

```diff
@@ -0,0 +1,309 @@
+
+(defn generate-resources
+  [number avg-resource-size title-size]
+  (let [tag-word-fn #(rnd/random-pronouncable-word 15 7 {:lowerb 5 :upperb 100})]
+    (map (fn [i]
+           (let [line-size 4 ;; stored as an int in the database...
+                 type-name (random-type)
+                 title (pseudonym "resource" i title-size)
+                 file (rnd/random-pp-path)
+                 resource-size (rnd/safe-sample-normal avg-resource-size (quot avg-resource-size 4))
+                 tags-mean (-> (quot resource-size 5) (min 200))
+                 tags-size (rnd/safe-sample-normal tags-mean (quot tags-mean 2) {:lowerb 10})
+                 parameters-size (max 0
+                                   (- resource-size
+                                      tags-size
+                                      (count type-name)
+                                      (count title)
+                                      (count file)
+                                      line-size))
+                 tags (build-to-size tags-size tag-word-fn)
+                 parameters (build-parameters parameters-size)
+                 resource (rnd/random-resource type-name title {"tags" tags "file" file})]
+             (assoc resource "parameters" parameters)))
+         (range number))))
+
+(defprotocol NamedEdges
+  (add [catalog r1 r2 relation] "Create an edge of relation between resource 1 and 2 in catalog."))
+
+(defrecord Catalog [graph edges]
+  ;; graph is a clojure.tools.namespace.dependency MapDependencyGraph of catalog resources.
+  ;; edges is a set tracking the relation type of each edge.
+  NamedEdges
+    (add [catalog r1 r2 relation]
+      (let [[source target] (case relation
+                               (:before :contains :notifies) [r1 r2]
+                               (:required-by :subscription-of) [r2 r1])]
+        (Catalog.
+          (dep/depend (:graph catalog) source target)
+          (conj (:edges catalog) {:source source :target target :relation relation})))))
+
+(defn new-catalog-graph
+  "Return a new initialized empty Catalog."
+  []
+  (->Catalog (dep/graph) #{}))
+
+(defn generate-catalog-graph
+  "Generate a set of containment edges associating each class with main-stage
+   and randomly distributing remaining resources within the classes. Returns
+   a Catalog instance.
+
+   Randomly generates additional required-by, notifies, before and
+   subscription-of between a given percentage of both class and non-class resources.
+
+   Makes use of Catalog to ensure we're still building a directed acyclic graph, and to
+   track the relation type between edges."
+  [main-stage classes resources additional-edge-percent]
+  (let [additional-class-edges (quot (* additional-edge-percent (count classes)) 100)
+        additional-resource-edges (quot (* additional-edge-percent (count resources)) 100)
+        shuffled-classes (set (shuffle classes))
```

Comment:
`set` appears to undo the shuffling process, because it all gets put into a hash.

Reply:
Yep, that doesn't work quite right. Thanks for catching that!. I've shifted it over so I only force it into a set before performing the cset/difference.